File: frontend/src/types/db/annotation.ts
import { z } from "zod";

const AnnotationSchema = z.object({
    id: z.number(),
    post_id: z.number(),
    product_id: z.number(),
    display_name: z.string(),
    x: z.number(),
    y: z.number(),
    created_at: z.string().datetime(),
    updated_at: z.string().datetime(),
});

type Annotation = z.infer<typeof AnnotationSchema>;

const ProductSchema = z.object({
    id: z.number(),
    name: z.string(),
    link: z.string(),
    created_at: z.string().datetime(),
    updated_at: z.string().datetime(),
});

type Product = z.infer<typeof ProductSchema>;

export { AnnotationSchema, ProductSchema };
export type { Annotation, Product };

File: backend/internal/domain/entity/annotation.go
package entity

import (
	"time"
)

type Annotation struct {
	ID          int64
	PostID      int64
	ProductID   *int64
	DisplayName string
	X           int
	Y           int
	CreatedAt   time.Time
	UpdatedAt   time.Time
}

func NewAnnotation(postID int64, productID *int64, displayName string, x, y int) *Annotation {
	now := time.Now()
	return &Annotation{
		PostID:      postID,
		ProductID:   productID,
		DisplayName: displayName,
		X:           x,
		Y:           y,
		CreatedAt:   now,
		UpdatedAt:   now,
	}
}